### PR TITLE
Improve vcgencmd

### PR DESCRIPTION
**Pull request by andrum99**
Pull in changes from orphaned PR #528: namely add basic usage, detect and report errors to STDERR, add link to forthcoming documentation on the website.

### CODE REVIEW

File: host_applications/linux/apps/gencmd/gencmd.c [hunk 1]
```diff
@@ -1,7 +1,6 @@
 /*
 Copyright (c) 2012, Broadcom Europe Ltd
 All rights reserved.
-
```
Comment (pelwell) [inline host_applications/linux/apps/gencmd/gencmd.c]: Why delete this blank lines?

File: host_applications/linux/apps/gencmd/gencmd.c [hunk 2]
```diff
@@ -36,41 +34,58 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
+   if ( argc == 1 )
    {
-       if (( strcmp( argv[1], "0" ) == 0 ) || ( strcmp( argv[1], "1" ) == 0 ))
-       {
-           instNum = atoi( argv[1] );
-           argv++;
-           argc--;
-       }
+      // no arguments passed, so show basic usage
+      show_usage();
```
Comment (pelwell) [inline host_applications/linux/apps/gencmd/gencmd.c]: I think you need a "return 0;" after this, to match the "-h" option.

File: host_applications/linux/apps/gencmd/gencmd.c [hunk 3]
```diff
@@ -36,41 +34,58 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 #include "interface/vmcs_host/vc_vchi_gencmd.h"
 
+void show_usage()
+{
+   printf( "Usage: vcgencmd [-t] command\n" );
+   printf( "Send a command to the VideoCore and print the result.\n\n" );
+   printf( "  -t         Time how long the command takes to complete\n" );
```
Comment (pelwell) [inline host_applications/linux/apps/gencmd/gencmd.c]: It is traditional (although a bit redundant) to include the '-h'/'--help' options in the list.

File: host_applications/linux/apps/gencmd/gencmd.c [hunk 4]
```diff
@@ -128,7 +143,19 @@ int main( int argc, char **argv )
...
+                fprintf (stderr, "%s\n", buffer);
+                if ( strcmp( buffer, "error=1 error_msg=\"Command not registered\"" ) == 0 )
+                {
+                   fprintf( stderr, "Use 'vcgencmd commands' to get a list of commands\n" );
+                }
+                return -2;
+             }
+             else
+             {
+               printf("%s\n", buffer );
```
Comment (pelwell) [inline host_applications/linux/apps/gencmd/gencmd.c]: What's wrong with `puts(buffer);`?

File: host_applications/linux/apps/gencmd/gencmd.c (hunk 1 above)
Comment (andrum99) [reply host_applications/linux/apps/gencmd/gencmd.c]: fixed

File: host_applications/linux/apps/gencmd/gencmd.c [hunk 5]
```diff
@@ -12,7 +12,6 @@ modification, are permitted provided that the following conditions are met:
     * Neither the name of the copyright holder nor the
       names of its contributors may be used to endorse or promote products
       derived from this software without specific prior written permission.
-
```
Comment (pelwell) [inline host_applications/linux/apps/gencmd/gencmd.c]: You missed this deletion.

File: host_applications/linux/apps/gencmd/gencmd.c [hunk 6]
```diff
@@ -36,41 +35,59 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 #include "interface/vmcs_host/vc_vchi_gencmd.h"
 
+void show_usage()
+{
+   printf( "Usage: vcgencmd [-t] command\n" );
+   printf( "Send a command to the VideoCore and print the result.\n\n" );
+   printf( "  -t          Time how long the command takes to complete\n" );
+   printf( "  -h, --help  Show this information" );
```
Comment (pelwell) [inline host_applications/linux/apps/gencmd/gencmd.c]: You've missed the newline, but these could all be puts() calls which wouldn't need the newline.

File: host_applications/linux/apps/gencmd/gencmd.c [hunk 7]
```diff
@@ -36,41 +36,60 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
         return -1;
     }
 
     vc_vchi_gencmd_init(vchi_instance, &vchi_connection, 1 );
 
     if (argc > 1)
     {
+       // first check if we were invoked with either -h or --help
+       // in which case show basic usage and exit
+      if( strcmp( argv[1], "-h" ) == 0 || strcmp( argv[1], "--help" ) == 0 )
```
Comment (pelwell) [inline host_applications/linux/apps/gencmd/gencmd.c]: Watch the indents here.

File: host_applications/linux/apps/gencmd/gencmd.c [hunk 8]
```diff
@@ -128,7 +147,19 @@ int main( int argc, char **argv )
...
+                fprintf (stderr, "%s\n", buffer);
+                if ( strcmp( buffer, "error=1 error_msg=\"Command not registered\"" ) == 0 )
+                {
+                   fprintf( stderr, "Use 'vcgencmd commands' to get a list of commands\n" );
+                }
+                return -2;
+             }
+             else
+             {
+               puts(buffer);
```
Comment (pelwell) [inline host_applications/linux/apps/gencmd/gencmd.c]: Indent.

File: host_applications/linux/apps/gencmd/gencmd.c [hunk 9]
```diff
@@ -36,41 +36,60 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
     if (argc > 1)
     {
+       // first check if we were invoked with either -h or --help
+       // in which case show basic usage and exit
+      if( strcmp( argv[1], "-h" ) == 0 || strcmp( argv[1], "--help" ) == 0 )
+      {
+         show_usage();
+         return 0;
+      }
+
```
Comment (pelwell) [inline host_applications/linux/apps/gencmd/gencmd.c]: There is some whitespace on this line - please delete it.

**Comment by andrum99** [comment]
> Indentation is all over the place in this file, but let's avoid adding new crimes against whitespace.

These are all cut and paste errors - I cut and pasted the whole file verbatim from the old PR, but I ran it via Notepad++ as I was going to edit something out. SOmewhere along the way it picked up contamination.

**Comment by andrum99** [comment]
OK, I've gone through the entire file by hand and fixed the indents. The file no longer uses spaces to indent lines - it is all tabs.

**Comment by pelwell** [comment]
Noooo! Userland should use 3 spaces (don't complain to me...) indents, absolutely no hard TABs. I wasn't asking you to change everything, just not make it any worse. My intention is to squash all of these commits into one, then at some later date we may have a separate, whitespace-only commit that fixes it all.

**Comment by andrum99** [comment]
No problem - Notepad++ can do that for me automagically.

**Comment by lurch** [comment]
@pelwell Might be worth adding an editorconfig? https://editorconfig.org/  :man_shrugging:

**Comment by lurch** [comment]
@andrum99 Something still looks screwy :confused:  https://github.com/raspberrypi/userland/pull/552/files

IMVHO a PR should always aim to minimise the differences, to make reviewing the _actual_ changes easier.

**Comment by JamesH65** [comment]
If you fancy it, "astyle -A1 -s3 -w" on the file(s) should bring it up to the VC4 coding standard, or something close, then submit another PR just for the whitespace changes (ie no functional changes). Will make future changes more consistent.

**Comment by andrum99** [comment]
> If you fancy it, "astyle -A1 -s3 -w" on the file(s) should bring it up to the VC4 coding standard, or something close, then submit another PR just for the whitespace changes (ie no functional changes). Will make future changes more consistent.

I ran that and it told me the file wan unchanged, so we've managed to do it the hard way already 😀 Thanks for getting this one across the line.

**Comment by andrum99** [comment]
> @andrum99 Something still looks screwy 😕 https://github.com/raspberrypi/userland/pull/552/files
> 
> IMVHO a PR should always aim to minimise the differences, to make reviewing the _actual_ changes easier.

There were no changes in this PR - it was a copy and paste from an abandoned one. I suspect all of the problems in this were also present in it.

**Comment by andrum99** [comment]
Please don't build a new debian package with this, until @JamesH65's documentation lands in the documentation repo.

**Pull request merged**